Remove hacked twins from codegen

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#36666 Remove hacked twins from codegen**

We need to introduce hacked twin overloads for ops taking lists of optional tensors.
I'm not really sure why actually, but them being a special case in codegen blocks removal of use_c10_dispatcher: unboxed_only.
This PR does not remove the "hacked twin" hack, but it removes it from codegen, instead manually specifying them in register_prim_ops.cpp and unblocking removal of use_c10_dispatcher: unboxed_only.

Note that GitHub shows many changes in register_prim_ops.cpp that actually don't exist, the GitHub diff view catches some code inbetween two far-away changed lines as changed too even though they didn't change. Check phabricator for a more accurate diff view.

Original commit changeset: c5e2386ad06a

Differential Revision: [D21044962](https://our.internmc.facebook.com/intern/diff/D21044962/)

smessmer committed:
Back out "Revert D20931186: Enable c10 unboxing for ops with TensorList"

Original commit changeset: c5e2386ad06a

Differential Revision: [D21044962](https://our.internmc.facebook.com/intern/diff/D21044962/)

[ghstack-poisoned]

smessmer committed:
Update on "Back out "Revert D20931186: Enable c10 unboxing for ops with TensorList""


Original commit changeset: c5e2386ad06a

Differential Revision: [D21044962](https://our.internmc.facebook.com/intern/diff/D21044962/)

[ghstack-poisoned]

smessmer committed:
Update on "Back out "Revert D20931186: Enable c10 unboxing for ops with TensorList""


Original commit changeset: c5e2386ad06a

Differential Revision: [D21044962](https://our.internmc.facebook.com/intern/diff/D21044962/)

[ghstack-poisoned]

smessmer committed:
Update on "Back out "Revert D20931186: Enable c10 unboxing for ops with TensorList""


Original commit changeset: c5e2386ad06a

Differential Revision: [D21044962](https://our.internmc.facebook.com/intern/diff/D21044962/)

[ghstack-poisoned]

smessmer committed:
Update on "Remove hacked twins from codegen"


We need to introduce hacked twin overloads for ops taking lists of optional tensors.
I'm not really sure why actually, but them being a special case in codegen blocks removal of use_c10_dispatcher: unboxed_only.
This PR does not remove the "hacked twin" hack, but it removes it from codegen, instead manually specifying them in register_prim_ops.cpp and unblocking removal of use_c10_dispatcher: unboxed_only.

Original commit changeset: c5e2386ad06a

Differential Revision: [D21044962](https://our.internmc.facebook.com/intern/diff/D21044962/)

[ghstack-poisoned]

smessmer committed:
Update on "Remove hacked twins from codegen"


We need to introduce hacked twin overloads for ops taking lists of optional tensors.
I'm not really sure why actually, but them being a special case in codegen blocks removal of use_c10_dispatcher: unboxed_only.
This PR does not remove the "hacked twin" hack, but it removes it from codegen, instead manually specifying them in register_prim_ops.cpp and unblocking removal of use_c10_dispatcher: unboxed_only.

Original commit changeset: c5e2386ad06a

Differential Revision: [D21044962](https://our.internmc.facebook.com/intern/diff/D21044962/)

[ghstack-poisoned]

smessmer committed:
Update on "Remove hacked twins from codegen"


We need to introduce hacked twin overloads for ops taking lists of optional tensors.
I'm not really sure why actually, but them being a special case in codegen blocks removal of use_c10_dispatcher: unboxed_only.
This PR does not remove the "hacked twin" hack, but it removes it from codegen, instead manually specifying them in register_prim_ops.cpp and unblocking removal of use_c10_dispatcher: unboxed_only.

Note that GitHub shows many changes in register_prim_ops.cpp that actually don't exist, the GitHub diff view catches some code inbetween two far-away changed lines as changed too even though they didn't change. Check phabricator for a more accurate diff view.

Original commit changeset: c5e2386ad06a

Differential Revision: [D21044962](https://our.internmc.facebook.com/intern/diff/D21044962/)

[ghstack-poisoned]

bhosmer reviewed (approved):
Will follow up to pull this out of prim ops. https://github.com/pytorch/pytorch/issues/36931

ezyang commented:
Hacked twins were introduced in #32574 by @bhosmer. So bhosmer should really weigh in here, but manually writing three overloads seems totally reasonable to me.

smessmer commented on torch/csrc/jit/runtime/register_prim_ops.cpp:
```diff
@@ -24,308 +24,376 @@ namespace jit {
+         [](Stack& stack) {
+           auto indices = pop(stack).toTensorVector();
+           auto self = pop(stack).toTensor();
+           auto result = at::index(std::move(self), std::move(indices));
```
hm I'll remove it for now but we should probably disable that warning, I think call sites ought to be allowed to use std::move even if the target function takes by reference, because it's not visible at the call site what the target function does and checking code correctness should be as local as possible.

ezyang replied:
I've mentioned this to you previously in code review: I don't think you should specify `std::move` unless it semantically makes sense to actually move into the function in question. I don't think you can just rely on people to write code where they won't continue to use the value after moving it, since it won't actually be moved out of.

smessmer committed:
Update on "Remove hacked twins from codegen"


We need to introduce hacked twin overloads for ops taking lists of optional tensors.
I'm not really sure why actually, but them being a special case in codegen blocks removal of use_c10_dispatcher: unboxed_only.
This PR does not remove the "hacked twin" hack, but it removes it from codegen, instead manually specifying them in register_prim_ops.cpp and unblocking removal of use_c10_dispatcher: unboxed_only.

Note that GitHub shows many changes in register_prim_ops.cpp that actually don't exist, the GitHub diff view catches some code inbetween two far-away changed lines as changed too even though they didn't change. Check phabricator for a more accurate diff view.

Original commit changeset: c5e2386ad06a

Differential Revision: [D21044962](https://our.internmc.facebook.com/intern/diff/D21044962/)

[ghstack-poisoned]

smessmer committed:
Update on "Remove hacked twins from codegen"


We need to introduce hacked twin overloads for ops taking lists of optional tensors.
I'm not really sure why actually, but them being a special case in codegen blocks removal of use_c10_dispatcher: unboxed_only.
This PR does not remove the "hacked twin" hack, but it removes it from codegen, instead manually specifying them in register_prim_ops.cpp and unblocking removal of use_c10_dispatcher: unboxed_only.

Note that GitHub shows many changes in register_prim_ops.cpp that actually don't exist, the GitHub diff view catches some code inbetween two far-away changed lines as changed too even though they didn't change. Check phabricator for a more accurate diff view.

Original commit changeset: c5e2386ad06a

Differential Revision: [D21044962](https://our.internmc.facebook.com/intern/diff/D21044962/)

[ghstack-poisoned]

smessmer committed:
Update on "Remove hacked twins from codegen"


We need to introduce hacked twin overloads for ops taking lists of optional tensors.
I'm not really sure why actually, but them being a special case in codegen blocks removal of use_c10_dispatcher: unboxed_only.
This PR does not remove the "hacked twin" hack, but it removes it from codegen, instead manually specifying them in register_prim_ops.cpp and unblocking removal of use_c10_dispatcher: unboxed_only.

Note that GitHub shows many changes in register_prim_ops.cpp that actually don't exist, the GitHub diff view catches some code inbetween two far-away changed lines as changed too even though they didn't change. Check phabricator for a more accurate diff view.

Original commit changeset: c5e2386ad06a

Differential Revision: [D21044962](https://our.internmc.facebook.com/intern/diff/D21044962/)

[ghstack-poisoned]
